### PR TITLE
Edit-message (8/n): Implement edit-message!

### PR DESCRIPTION
**Pull request by chrisbobbe**
Edit-message! 🎉

Fixes: #126

## Screenshots

| Light | Dark |
| --- | --- |
| image | image |
| image | image |
| image | image |
| image | image |
| image | image |

### CODE REVIEW

**Comment by chrisbobbe** [comment]
Oh, and if you tap "Edit message" in the action sheet but there's already text in the compose box, here's the dialog you get. Giving the "Discard" button a destructive/red style is future work, #1032.

| Light | Dark |
| --- | --- |
| image | image |

**Comment by alya** [comment]
Did we discuss the strings anywhere? I would maybe write something like this:

> Discard the message you're writing?
> When you edit a message, the content that was previously in the compose box is discarded.
[ Cancel] [Discard content and continue]

**Comment by alya** [comment]
I think it's good to put a `.` at the end of user-visible error strings.

**Comment by chrisbobbe** [comment]
Makes sense; I started a discussion on CZO: [#mobile-team > Edit-message: UI strings @ 💬](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/Edit-message.3A.20UI.20strings/near/2168048)

**Comment by chrisbobbe** [comment]
Revision pushed, with those UI string updates, and I've updated the screenshots.

**Comment by chrisbobbe** [comment]
(Oops, fixed a failing test.)

**Review by PIG208 — COMMENTED** [review]
Thanks for the PR! Looks pretty good to me overall. Left some comments.

File: lib/widgets/message_list.dart [hunk 1]
```diff
@@ -1462,10 +1533,11 @@ class MessageWithPossibleSender extends StatelessWidget {
               Expanded(child: Column(
                 crossAxisAlignment: CrossAxisAlignment.stretch,
                 children: [
-                  MessageContent(message: message, content: item.content),
+                  content,
                   if ((message.reactions?.total ?? 0) > 0)
                     ReactionChipsList(messageId: message.id, reactions: message.reactions!),
-                  if (editStateText != null)
+                  if (editMessageErrorStatusRow != null) editMessageErrorStatusRow
```
Comment (gnprice) [inline lib/widgets/message_list.dart:1539]: I think this would go well as its own widget:
```suggestion
                  if (editMessageErrorStatus != null)
                    _EditMessageStatusRow(messageId: message.id),
```

That'd help keep this build method mostly clear of details that are specific to a message currently being edited.

File: lib/widgets/message_list.dart [hunk 2]
```diff
@@ -1445,9 +1458,67 @@ class MessageWithPossibleSender extends StatelessWidget {
...
+      };
+    }
+
+    Widget content = MessageContent(message: message, content: item.content);
+
+    if (editMessageErrorStatus != null) {
+      content = Opacity(opacity: 0.6, child: content);
+      switch (editMessageErrorStatus) {
+        case true:
+          content = _withRestoreEditMessageGestureDetector(context, content);
```
Comment (gnprice) [inline lib/widgets/message_list.dart:1502]: This also seems like a good candidate for its own widget:
```suggestion
          content = _RestoreEditMessageGestureDetector(messageId: item.message.id, child: content);
```

File: lib/widgets/message_list.dart [hunk 3]
```diff
@@ -1445,9 +1458,67 @@ class MessageWithPossibleSender extends StatelessWidget {
...
+
     return GestureDetector(
       behavior: HitTestBehavior.translucent,
-      onLongPress: () => showMessageActionSheet(context: context, message: message),
+      onLongPress: () {
+        final composeBoxState = MessageListPage.ancestorOf(context).composeBoxState;
+        final controller = composeBoxState?.controller;
+        final editMessageInProgress =
+          controller is EditMessageComposeBoxController
+          || editMessageErrorStatus != null;
```
Comment (gnprice) [inline lib/widgets/message_list.dart:1515]: Can this logic all move into `showMessageActionSheet`? Seems like between `context` and `message` it has all the inputs it needs in order to do the same calculations.

**Review by gnprice — COMMENTED** [review]
Exciting! A few quick comments from an early skim.

File: lib/widgets/message_list.dart [hunk 4]
```diff
@@ -1445,9 +1458,67 @@ class MessageWithPossibleSender extends StatelessWidget {
...
+
+    Widget content = MessageContent(message: message, content: item.content);
+
+    if (editMessageErrorStatus != null) {
+      content = Opacity(opacity: 0.6, child: content);
+      switch (editMessageErrorStatus) {
+        case true:
+          content = _withRestoreEditMessageGestureDetector(context, content);
+        case false:
+          content = IgnorePointer(child: content);
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply lib/widgets/message_list.dart:1504]: It's meant to neutralize interactive message content like links, spoilers, etc.; I thought it made sense along with the faded 0.6-opacity appearance in the Figma. I can add a comment about that.

File: lib/widgets/action_sheet.dart [hunk 5]
```diff
@@ -568,6 +574,18 @@ void showMessageActionSheet({required BuildContext context, required Message mes
   final markAsUnreadSupported = store.zulipFeatureLevel >= 155; // TODO(server-6)
   final showMarkAsUnreadButton = markAsUnreadSupported && isMessageRead;
 
+  final now = ZulipBinding.instance.utcNow().millisecondsSinceEpoch ~/ 1000;
+  final editLimit = store.realmMessageContentEditLimitSeconds;
+  final outsideEditLimit =
+    editLimit != null
+    && editLimit != 0 // TODO(server-6) remove (pre-FL 138, 0 represents no limit)
+    && now - message.timestamp > editLimit;
+  final showEditButton = message.senderId == store.selfUserId
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply lib/widgets/action_sheet.dart:583]: Oh, TIL; thanks.

File: lib/widgets/message_list.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply lib/widgets/message_list.dart:1515]: This condition means whether the compose box is in edit-message mode instead of the default write-new-message mode.

File: lib/widgets/action_sheet.dart [hunk 6]
```diff
@@ -568,6 +574,18 @@ void showMessageActionSheet({required BuildContext context, required Message mes
...
+    && messageListHasComposeBox
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply lib/widgets/action_sheet.dart:584]: Oh indeed, I didn't see that we already have the value we need.

File: lib/widgets/message_list.dart [hunk 7]
```diff
@@ -1445,9 +1458,67 @@ class MessageWithPossibleSender extends StatelessWidget {
...
+            style: baseTextStyle.copyWith(color: designVariables.btnLabelAttLowIntDanger),
+            textAlign: TextAlign.end,
+            zulipLocalizations.savingMessageEditFailedLabel)),
+      };
+    }
+
+    Widget content = MessageContent(message: message, content: item.content);
+
+    if (editMessageErrorStatus != null) {
+      content = Opacity(opacity: 0.6, child: content);
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply lib/widgets/message_list.dart:1499]: Left a comment about this after asking Greg in the office:

```
      // The Figma also fades the sender row:
      //   https://github.com/zulip/zulip-flutter/pull/1498#discussion_r2076574000
      // We've decided to just fade the message content because that's the only
      // thing that's changing.
```

File: test/widgets/action_sheet_test.dart [hunk 8]
```diff
@@ -1410,6 +1418,174 @@ void main() {
       });
     });
 
+    group('EditButton', () {
+      group('present/absent appropriately', () {
+        /// Test whether the edit-message button is visible, given params.
+        ///
+        /// The message timestamp is 60s before the current time
+        /// ([TestBinding.utcNow]) as of the start of the test run.
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply test/widgets/action_sheet_test.dart:1426]: Indeed—also, thanks for creating this :)

File: test/widgets/action_sheet_test.dart [hunk 9]
```diff
@@ -1410,6 +1418,174 @@ void main() {
...
+              timestamp: eg.utcTimestamp(testBinding.utcNow()) - 60);
+
+            await setupToMessageActionSheet(tester,
+              message: message,
+              narrow: narrow ?? TopicNarrow.ofMessage(message),
+              realmAllowMessageEditing: allowed,
+              realmMessageContentEditLimitSeconds: limit,
+            );
+
+            if (!composing && errorStatus == null) {
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply test/widgets/action_sheet_test.dart:1483]: I've added some comments to try to clarify.

**Comment by chrisbobbe** [comment]
Thanks for the reviews! Revision pushed, for Zixuan's review.

**Comment by PIG208** [comment]
Looks like there are some compilation errors in lib/widgets/message_list.dart, probably from some non-local merge conflicts.

File: test/widgets/action_sheet_test.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (PIG208) [reply test/widgets/action_sheet_test.dart]: Glad that helps!

File: test/widgets/action_sheet_test.dart [hunk 10]
```diff
@@ -1410,6 +1429,196 @@ void main() {
...
+          bool allowed = true,
+          int? limit,
+          bool composing = false,
+          bool? errorStatus,
+          bool poll = false,
+        }) {
+          // It's impossible for the compose box to be in editing mode
+          // while, for the same message, the message-edit request is either
+          // in progress or in the error state.
+          assert(!composing || errorStatus == null);
```
Comment (PIG208) [inline test/widgets/action_sheet_test.dart:1462]: Oh I see! Yeah this clears up the meaning of `composing`.

I think it is still possible if you initiate the edit in both the channel narrow and the topic narrow containing it, then save the edit in the topic narrow, navigate back and save in the channel narrow, on the same message. If the first request takes long enough to finish. But that's quite an edge case, and is guarded by a containsKey check already.

**Comment by PIG208** [comment]
Thanks for the update! This looks good to me other than the compilation errors.

File: lib/widgets/compose_box.dart [hunk 11]
```diff
@@ -1484,6 +1552,60 @@ class _ErrorBanner extends _Banner {
...
+    final zulipLocalizations = ZulipLocalizations.of(context);
+    return Row(mainAxisSize: MainAxisSize.min, spacing: 8, children: [
+      ZulipWebUiKitButton(label: zulipLocalizations.composeBoxBannerButtonCancel,
+        onPressed: () {
+          composeBoxState.endEditInteraction();
+        }),
+      // TODO(#1481) disabled appearance when there are validation errors
+      ZulipWebUiKitButton(label: zulipLocalizations.composeBoxBannerButtonSave,
+        attention: ZulipWebUiKitButtonAttention.high,
+        onPressed: () {
```
Comment (gnprice) [inline lib/widgets/compose_box.dart:1584]: nit: this callback is about 20 lines; once it's more than a small handful of lines, I think it's a bit easier to read as a separate named function, so here a private method on this class.

File: lib/widgets/message_list.dart [hunk 12]
```diff
@@ -1446,6 +1446,26 @@ class MessageWithPossibleSender extends StatelessWidget {
...
+
+    final editMessageErrorStatus = store.getEditMessageErrorStatus(message.id);
+    if (editMessageErrorStatus != null) {
+      // The Figma also fades the sender row:
+      //   https://github.com/zulip/zulip-flutter/pull/1498#discussion_r2076574000
+      // We've decided to just fade the message content because that's the only
+      // thing that's changing.
+      content = Opacity(opacity: 0.6, child: content);
+      switch (editMessageErrorStatus) {
+        case false:
```
Comment (gnprice) [inline lib/widgets/message_list.dart:1459]: nit: How about `if (!editMessageErrorStatus)`?

That'll have basically the same exhaustiveness-checking effect, too — `if` requires a bool and will reject a `bool?`.

File: lib/widgets/message_list.dart [hunk 13]
```diff
@@ -1481,3 +1503,77 @@ class MessageWithPossibleSender extends StatelessWidget {
...
+        spacing: 1.5,
+        children: [
+          Text(
+            style: baseTextStyle
+              .copyWith(color: designVariables.btnLabelAttLowIntInfo),
+            textAlign: TextAlign.end,
+            zulipLocalizations.savingMessageEditLabel),
+          // TODO instead place within outer padding; see Figma:
+          //   https://www.figma.com/design/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=4026-8775&m=dev
+          LinearProgressIndicator(
```
Comment (gnprice) [inline lib/widgets/message_list.dart:1540]: I don't quite follow this comment. Which padding should this be placed within?

File: lib/widgets/message_list.dart [hunk 14]
```diff
@@ -1481,3 +1503,77 @@ class MessageWithPossibleSender extends StatelessWidget {
...
+  @override
+  Widget build(BuildContext context) {
+    return GestureDetector(
+      behavior: HitTestBehavior.opaque,
+      onTap: () {
+        final store = PerAccountStoreWidget.of(context);
+        final composeBoxState = MessageListPage.ancestorOf(context).composeBoxState;
+        if (composeBoxState == null) return;
+        composeBoxState.startEditInteraction(messageId: messageId,
+          originalRawContent: store.takeFailedMessageEdit(messageId));
```
Comment (gnprice) [inline lib/widgets/message_list.dart:1575]: Hmm. This `originalRawContent` value is the right value to initialize the text input to — but it's not the right value to use when sending the resulting edit request to the server. For that, we want the *original* original raw content, because that's the value the user is (implicitly) expecting the server will have.

Probably the right fix is to add another field on `_EditMessageRequestStatus`.

File: test/widgets/action_sheet_test.dart (hunk 10 above)
Comment (chrisbobbe) [reply test/widgets/action_sheet_test.dart:1462]: Oh, good find! I can address this edge case in my next revision.

File: assets/l10n/app_en.arb [hunk 15]
```diff
@@ -367,6 +411,10 @@
       "destination": {"type": "String", "example": "#channel name > topic name"}
     }
   },
+  "composeBoxEditMessageHint": "Message content",
+  "@composeBoxEditMessageHint": {
+    "description": "Hint text for content input when editing a message."
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply assets/l10n/app_en.arb:416]: Sure; going with no hint text.

File: lib/widgets/message_list.dart (hunk 13 above)
Comment (chrisbobbe) [reply lib/widgets/message_list.dart:1540]: Here are screenshots of the Figma with and without that outer padding highlighted:

image

image

The progress indicator sits inside that outer padding.

I think it looks fine as-is and would be OK to remove the TODO, revisiting if needed in the future.

File: lib/widgets/message_list.dart (hunk 14 above)
Comment (chrisbobbe) [reply lib/widgets/message_list.dart:1575]: I see; yeah, thanks for catching this.

File: lib/widgets/message_list.dart (hunk 13 above)
Comment (gnprice) [reply lib/widgets/message_list.dart:1540]: Hmm, those screenshots don't really help. :slightly_smiling_face: How does this version differ from what's in Figma?

Looking into it a bit more, I think you're referring to the 4px of bottom padding on the MessageWithPossibleSender.

So in the Figma design, the indicator encroaches on that padding and gets closer to whatever's below (like the next message, or a date separator). In this version, when the indicator appears, it instead makes the message taller and stays 4px away from the sibling below.

I see the appeal of the version in Figma, but the difference is subtle. Happy to leave it as a TODO.

File: lib/widgets/message_list.dart (hunk 14 above)
Comment (chrisbobbe) [reply lib/widgets/message_list.dart:1575]: I've had some more thoughts prompted by this—will share later today

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

As discussed in the office on Friday, this makes some improvements over the last revision—but omits the refetch-raw-content request on restoring a failed edit, which I'd been considering. 🙂 (Omits it because the edge case it addresses is minuscule, and it adds time and risk of failure in bad network connections.)

In particular:
- Now if you tap to restore a failed edit but there's already text in the compose box, and you cancel the "Discard…?" confirmation dialog, the failed edit hasn't been consumed.
- The edit-message compose box has a disabled "Preparing…" state for when the fetch-raw-content request is in progress (which happens just when you start an edit from the action sheet).

Here are screenshots of that "Preparing…" state, with the normal state for comparison:

| Dark | Light |
| --- | --- |
| image | image |
| image | image |

Giving the "Save" button a disabled appearance, in this state, is #1481; we don't have a design for that.

File: lib/widgets/compose_box.dart [hunk 16]
```diff
@@ -1515,12 +1664,141 @@ class ComposeBox extends StatefulWidget {
...
+      return;
+    }
+    // TODO scroll message list to ensure the message is still in view;
+    //   highlight it?
+    setState(() {
+      assert(_controller is EditMessageComposeBoxController);
+      (_controller as EditMessageComposeBoxController)
+        ..originalRawContent = fetchedRawContent
+        ..content.value = TextEditingValue(text: fetchedRawContent)
+        ..contentFocusNode.requestFocus();
```
Comment (gnprice) [inline lib/widgets/compose_box.dart:1786]: This step doesn't seem to work for me empirically (on my Android 15 phone). When I hit "Edit message" on the message action sheet, the input doesn't get focused — I have to tap it to focus (and to cause the keyboard to appear).

OTOH the corresponding step in the `_byRestoringFailedEdit` case *does* work.

Do you reproduce?

This is a perfectly acceptable glitch for launch: it just means an extra tap, and I think users are accustomed to having to sometimes tap an input field to get the keyboard to open when it seems like it should happen automatically. So let's only spend a small amount of effort now on diagnosing or fixing, and if that isn't enough then just file a short issue to record the point for post-launch.

File: lib/widgets/compose_box.dart [hunk 17]
```diff
@@ -1515,12 +1664,141 @@ class ComposeBox extends StatefulWidget {
...
+      case false:
+        // This can happen if you start an edit interaction on one
+        // MessageListPage and then do an edit on a different MessageListPage,
+        // and the second edit is still saving when you return to the first.
+        //
+        // Abort rather than sending a request with a prevContentSha256
+        // that the server might not accept, and don't clear the compose
+        // box, so the user can try again after the request settles.
+        // TODO write a test for this
+        showErrorDialog(context: context,
```
Comment (gnprice) [inline lib/widgets/compose_box.dart:1711]: Yeah, I'm OK with skipping a test for this case.

**Review by gnprice — COMMENTED** [review]
Thanks for the revision! This behavior all now looks good, except one glitch mentioned below which I'm fine with for launch.

I haven't yet read the tests in the last/main commit.

File: lib/widgets/compose_box.dart (hunk 16 above)
Comment (chrisbobbe) [reply lib/widgets/compose_box.dart:1786]: Debugged and fixed! 🎉

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

File: lib/widgets/compose_box.dart [hunk 18]
```diff
@@ -1675,12 +1825,146 @@ class ComposeBox extends StatefulWidget {
...
+  void startEditInteraction(int messageId) async {
+    if (await _abortBecauseContentInputNotEmpty()) return;
+    if (!mounted) return;
+
+    final store = PerAccountStoreWidget.of(context);
+    final zulipLocalizations = ZulipLocalizations.of(context);
+
+    switch (store.getEditMessageErrorStatus(messageId)) {
+      case null:
+        _editFromRawContentFetch(messageId);
```
Comment (gnprice) [inline lib/widgets/compose_box.dart:1860]: bump https://github.com/zulip/zulip-flutter/pull/1498#discussion_r2092232833: put the two non-error cases next to each other

File: lib/widgets/compose_box.dart (hunk 16 above)
Comment (gnprice) [reply lib/widgets/compose_box.dart]: (FTR, the fix was the `addPostFrameCallback` here, for the reason its comment explains:
```dart
    setState(() {
      // setState to refresh the input, upload buttons, etc.
      // out of the disabled "Preparing…" state.
      editMessageController.originalRawContent = fetchedRawContent;
    });
    editMessageController.content.value = TextEditingValue(text: fetchedRawContent);
    SchedulerBinding.instance.addPostFrameCallback((_) {
      // post-frame callback so this happens after the input is enabled
      editMessageController.contentFocusNode.requestFocus();
    });
```
)

**Review by gnprice — COMMENTED** [review]
Thanks! Glad that glitch turned out to have a nice clean fix.

All looks good now except one nit — oh and I should read the tests :slightly_smiling_face:

File: test/widgets/compose_box_test.dart [hunk 19]
```diff
@@ -1354,4 +1400,462 @@ void main() {
...
+        checkNotInEditingMode(tester, narrow: narrow);
+      });
+    }
+    doTestInterruptComposingFromActionSheet(narrow: channelNarrow);
+    doTestInterruptComposingFromActionSheet(narrow: topicNarrow);
+    doTestInterruptComposingFromActionSheet(narrow: dmNarrow);
+
+    // Test the "Discard…?" confirmation dialog when there's text in the compose
+    // box for a new message, but you've said you want to edit a message.
+    void doTestInterruptComposingFromFailedEdit({required Narrow narrow}) {
```
Comment (gnprice) [inline test/widgets/compose_box_test.dart:1666]: This comment doesn't seem to match the name. (The comment seems to say the same thing as the comment on doTestInterruptComposingFromActionSheet above.)

File: test/widgets/compose_box_test.dart [hunk 20]
```diff
@@ -1354,4 +1400,462 @@ void main() {
...
+        checkNotInEditingMode(tester, narrow: narrow);
+
+        // …or send a new message.
+        connection.prepare(json: {}); // for typing-start request
+        connection.prepare(json: {}); // for typing-stop request
+        await enterContent(tester, 'new message to send');
+        state.controller.contentFocusNode.unfocus();
+        await tester.pump();
+        check(connection.takeRequests()).deepEquals([
+          (Subject<Object?> it) => it.isA<http.Request>()
```
Comment (gnprice) [inline test/widgets/compose_box_test.dart:1814]: nit: this sort of `deepEquals` argument is a bit shorter using `Condition`:
```suggestion
        check(connection.takeRequests()).deepEquals(<Condition<Object?>>[
          (it) => it.isA<http.Request>()
```

File: test/widgets/compose_box_test.dart [hunk 21]
```diff
@@ -1354,4 +1400,462 @@ void main() {
...
+
+        // Save; check that the request is made and the compose box resets.
+        connection.prepare(json: UpdateMessageResult().toJson());
+        await tester.tap(find.widgetWithText(ZulipWebUiKitButton, 'Save'));
+        checkRequest(messageId, prevContent: 'foo', content: 'baz');
+        await tester.pump(Duration.zero);
+        checkNotInEditingMode(tester, narrow: narrow);
+      });
+    }
+    doTestInterruptComposingFromFailedEdit(narrow: channelNarrow);
```
Comment (gnprice) [inline test/widgets/compose_box_test.dart:1714]: nit: for helpers that define a test case like `test` and `testWidgets` do, name them `testFoo` to match those:
```suggestion
    testInterruptComposingFromFailedEdit(narrow: channelNarrow);
```

We're mostly consistent about that practice already; for examples, see `git grep -P '\btest[^W]\w+\('`. (Or refine the pattern to possibly find more examples; it's excluding `testWidgets` but crudely so.)

**Review by gnprice — COMMENTED** [review]
OK, and here's a review of the remaining portion: the tests. They look good! Only small comments below.

These are definitely thorough. :sweat_smile: It's good to know that we can write such thorough tests of complex UI flows when we need to, and that we can do so while keeping them readable as these tests are. In general, though, I think our balance should be more in the direction of completing features faster, by spending less effort on comprehensive tests, than this represents.

One other aspect of these tests is that they have a noticeable runtime cost when running this test file. Here's timings on my office desktop (which is somewhat less speedy than my home desktop) with and without the `edit message` test group (knocking it out by inserting a `return;`):
```
$ time flutter test test/widgets/compose_box_test.dart 
00:17 +116: All tests passed!                                                              

time: 19.781s wall (25.325s u, 3.213s s)


$ time flutter test test/widgets/compose_box_test.dart 
00:10 +92: All tests passed!                                                               

time: 13.379s wall (16.966s u, 2.759s s)
```

So the 24 new test cases add about 6 seconds, or 50%, to the time to run the test file. Hardly a catastrophe, but enough to be a noticeable drag in the future when working on unrelated tests in this file (or unrelated changes to the compose box) and so repeatedly running this file.

I think one contributor to that runtime is that there's the beginnings of a combinatorial explosion: each test scenario gets run under three different narrows, including for interactions that seem unlikely to behave differently depending on the narrow.

Are there any of these where you could delete the `doTestFoo` calls for some of the narrows and it'd feel like all the same codepaths are effectively tested? If so, it'd be worth doing that.

If not, then the next point might be too much change to be worth doing before merging this, but would be good to have in mind for future tests: some of the additional checks like these, within some of the `doTestFoo` functions:
```dart
        // Now that we have the raw content, check the input is interactive
        // but no typing notifications are sent…
        // […]
        await enterContent(tester, 'some new content');
        // […]

        // …and the upload buttons work.
        // […]
        await tester.tap(find.byIcon(ZulipIcons.attach_file), warnIfMissed: false);
```
seem like they'd be just as effective if they were run for only one narrow, and only one value of other parameters like `start`. Separating them out is basically a form of making test cases more focused; and among the benefits of that sort of separation is that it can help a cluster of test cases end up doing N + M work, instead of N * M, while still effectively covering all the logic.

File: lib/widgets/compose_box.dart (hunk 18 above)
Comment (chrisbobbe) [reply lib/widgets/compose_box.dart:1860]: Ah! I put `null` and `false` together as the "two non-error cases", thinking of "error" as it's used in `getEditMessageErrorStatus` which produces this value:

```dart
  @override
  bool? getEditMessageErrorStatus(int messageId) =>
    _editMessageRequests[messageId]?.hasError;
```

But I see now what you mean, that the error case is the one where we show an error dialog. 😛

File: test/widgets/action_sheet_test.dart [hunk 22]
```diff
@@ -1421,6 +1440,174 @@ void main() {
...
+              // We're testing the request-in-progress state. Prepare a delay
+              // and tap Save.
+              connection.prepare(apiException: eg.apiBadRequest(), delay: Duration(seconds: 1));
+              await tester.tap(find.widgetWithText(ZulipWebUiKitButton, 'Save'));
+              await tester.pump(Duration.zero);
+            } else if (errorStatus == false) {
+              // We're testing the request-failed state. Prepare a failure
+              // and tap Save.
+              connection.prepare(
+                json: UpdateMessageResult().toJson(), delay: Duration(seconds: 1));
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply test/widgets/action_sheet_test.dart:1543]: Eep, thanks!

**Comment by gnprice** [comment]
Thanks for the revision! Merging, with a couple of tweaks in those tests:
```diff
--- test/widgets/compose_box_test.dart
+++ test/widgets/compose_box_test.dart
@@ -1538,3 +1538,3 @@ void main() {
 
-    void smokeTest({required Narrow narrow, required _EditInteractionStart start}) {
+    void testSmoke({required Narrow narrow, required _EditInteractionStart start}) {
       testWidgets('smoke: $narrow, ${start.message()}', (tester) async {
@@ -1579,2 +1579,3 @@ void main() {
         // TODO could also check that quote-and-reply and autocomplete work
+        //   (but as their own test cases, for a single narrow and start)
 
@@ -1589,8 +1590,8 @@ void main() {
     }
-    smokeTest(narrow: channelNarrow, start: _EditInteractionStart.actionSheet);
-    smokeTest(narrow: topicNarrow,   start: _EditInteractionStart.actionSheet);
-    smokeTest(narrow: dmNarrow,      start: _EditInteractionStart.actionSheet);
-    smokeTest(narrow: channelNarrow, start: _EditInteractionStart.restoreFailedEdit);
-    smokeTest(narrow: topicNarrow,   start: _EditInteractionStart.restoreFailedEdit);
-    smokeTest(narrow: dmNarrow,      start: _EditInteractionStart.restoreFailedEdit);
+    testSmoke(narrow: channelNarrow, start: _EditInteractionStart.actionSheet);
+    testSmoke(narrow: topicNarrow,   start: _EditInteractionStart.actionSheet);
+    testSmoke(narrow: dmNarrow,      start: _EditInteractionStart.actionSheet);
+    testSmoke(narrow: channelNarrow, start: _EditInteractionStart.restoreFailedEdit);
+    testSmoke(narrow: topicNarrow,   start: _EditInteractionStart.restoreFailedEdit);
+    testSmoke(narrow: dmNarrow,      start: _EditInteractionStart.restoreFailedEdit);
 
```

So being complete about that pattern from https://github.com/zulip/zulip-flutter/pull/1498#discussion_r2093814509; and making sure that TODO doesn't point someone in the wrong direction with respect to https://github.com/zulip/zulip-flutter/pull/1498#issuecomment-2892069490 .
